File: plugins/plugin-elizacloud/typescript/services/cloud-auth.ts
/**
 * CloudAuthService — Device-based auto-signup and session management.
 *
 * On first launch, derives a hardware fingerprint and calls
 * POST /api/v1/device-auth. The cloud backend creates a user + org +
 * $5 credit + API key if new, or returns the existing session.
 */

import { type IAgentRuntime, logger, Service } from "@elizaos/core";
import { CloudApiClient } from "../utils/cloud-api";
import type {
  CloudCredentials,
  DeviceAuthResponse,
  DevicePlatform,
} from "../types/cloud";
import { DEFAULT_CLOUD_CONFIG } from "../types/cloud";

/** SHA-256 hash of hostname + platform + arch + cpu + memory. */
async function deriveDeviceId(): Promise<string> {
  const os = await import("node:os");
  const crypto = await import("node:crypto");
  const cpus = os.cpus();
  const raw = [os.hostname(), os.platform(), os.arch(), cpus[0]?.model ?? "?", cpus.length, os.totalmem()].join(":");
  return crypto.createHash("sha256").update(raw).digest("hex");
}

function detectPlatform(): DevicePlatform {
  if (typeof process === "undefined") return "web";
  const map: Record<string, DevicePlatform> = { darwin: "macos", win32: "windows", linux: "linux" };
  return map[process.platform] ?? "linux";
}

export class CloudAuthService extends Service {
  static serviceType = "CLOUD_AUTH";
  capabilityDescription = "ElizaCloud device authentication and session management";

  private client: CloudApiClient;
  private credentials: CloudCredentials | null = null;

  constructor(runtime?: IAgentRuntime) {
    super(runtime);
    this.client = new CloudApiClient(DEFAULT_CLOUD_CONFIG.baseUrl);
  }

  static async start(runtime: IAgentRuntime): Promise<Service> {
    const service = new CloudAuthService(runtime);
    await service.initialize();
    return service;
  }

  async stop(): Promise<void> {
    this.credentials = null;
  }

  private async initialize(): Promise<void> {
    const baseUrl = String(this.runtime.getSetting("ELIZAOS_CLOUD_BASE_URL") ?? DEFAULT_CLOUD_CONFIG.baseUrl);
    this.client.setBaseUrl(baseUrl);

    // Try existing API key first
    const existingKey = this.runtime.getSetting("ELIZAOS_CLOUD_API_KEY");
    if (existingKey) {
      const key = String(existingKey);
      this.client.setApiKey(key);
      const valid = await this.validateApiKey(key);
      if (valid) {
        this.credentials = {
          apiKey: key,
          userId: String(this.runtime.getSetting("ELIZAOS_CLOUD_USER_ID") ?? ""),
          organizationId: String(this.runtime.getSetting("ELIZAOS_CLOUD_ORG_ID") ?? ""),
          authenticatedAt: Date.now(),
        };
        logger.info("[CloudAuth] Authenticated with existing API key");
        return;
      }
      logger.warn("[CloudAuth] Existing API key invalid, attempting device auth");
    }

    // Device-based auto-signup when explicitly enabled
    const enabled = this.runtime.getSetting("ELIZAOS_CLOUD_ENABLED");
    if (enabled === "true" || enabled === "1") {
      await this.authenticateWithDevice();
    } else {
      logger.info("[CloudAuth] Cloud not enabled (set ELIZAOS_CLOUD_ENABLED=true)");
    }
  }

  private async validateApiKey(key: string): Promise<boolean> {
    const resp = await fetch(`${this.client.getBaseUrl()}/models`, {
      headers: { Authorization: `Bearer ${key}` },
    });
    return resp.ok;
  }

  async authenticateWithDevice(): Promise<CloudCredentials> {
    const deviceId = await deriveDeviceId();
    const platform = detectPlatform();
    const appVersion = process.env.ELIZAOS_CLOUD_APP_VERSION ?? "2.0.0-alpha";
    const os = await import("node:os");

    logger.info(`[CloudAuth] Authenticating device (platform=${platform})`);

    const response = await this.client.postUnauthenticated<DeviceAuthResponse>("/device-auth", {
      deviceId,
      platform,
      appVersion,
      deviceName: os.hostname(),
    });

    this.credentials = {
      apiKey: response.data.apiKey,
      userId: response.data.userId,
      organizationId: response.data.organizationId,
      authenticatedAt: Date.now(),
    };
    this.client.setApiKey(response.data.apiKey);

    const action = response.data.isNew ? "New account created" : "Authenticated";
    logger.info(`[CloudAuth] ${action} (credits: $${response.data.credits.toFixed(2)})`);

    return this.credentials;
  }

  isAuthenticated(): boolean { return this.credentials !== null; }
  getCredentials(): CloudCredentials | null { return this.credentials; }
  getApiKey(): string | undefined { return this.credentials?.apiKey ?? this.client.getApiKey(); }
  getClient(): CloudApiClient { return this.client; }
  getUserId(): string | undefined { return this.credentials?.userId; }
  getOrganizationId(): string | undefined { return this.credentials?.organizationId; }
}

File: plugins/plugin-elizacloud/rust/src/services/cloud_auth.rs
//! CloudAuthService — Device-based auto-signup and session management.

use std::collections::HashMap;
use tracing::{info, warn};

use crate::cloud_api::CloudApiClient;
use crate::cloud_types::{CloudCredentials, CloudPluginConfig};
use crate::error::{ElizaCloudError, Result};

/// ElizaCloud device authentication and session management.
pub struct CloudAuthService {
    client: CloudApiClient,
    credentials: Option<CloudCredentials>,
}

impl CloudAuthService {
    /// Create a new auth service with default configuration.
    pub fn new() -> Result<Self> {
        let config = CloudPluginConfig::default();
        let client = CloudApiClient::new(&config.base_url, None)?;
        Ok(Self {
            client,
            credentials: None,
        })
    }

    /// Initialize with runtime settings.
    pub async fn start(&mut self, settings: &HashMap<String, String>) -> Result<()> {
        let base_url = settings
            .get("ELIZAOS_CLOUD_BASE_URL")
            .map(String::as_str)
            .unwrap_or("https://www.elizacloud.ai/api/v1");
        self.client.set_base_url(base_url);

        // Try existing API key
        if let Some(key) = settings.get("ELIZAOS_CLOUD_API_KEY") {
            self.client.set_api_key(key);
            if self.validate_api_key(key).await {
                self.credentials = Some(CloudCredentials {
                    api_key: key.clone(),
                    user_id: settings.get("ELIZAOS_CLOUD_USER_ID").cloned().unwrap_or_default(),
                    organization_id: settings.get("ELIZAOS_CLOUD_ORG_ID").cloned().unwrap_or_default(),
                    authenticated_at: std::time::SystemTime::now()
                        .duration_since(std::time::UNIX_EPOCH)
                        .unwrap_or_default()
                        .as_secs_f64(),
                });
                info!("[CloudAuth] Authenticated with existing API key");
                return Ok(());
            }
            warn!("[CloudAuth] Existing API key invalid");
        }

        let enabled = settings.get("ELIZAOS_CLOUD_ENABLED").map(String::as_str);
        if enabled == Some("true") || enabled == Some("1") {
            self.authenticate_with_device().await?;
        } else {
            info!("[CloudAuth] Cloud not enabled (set ELIZAOS_CLOUD_ENABLED=true)");
        }

        Ok(())
    }

    /// Stop the service and clear credentials.
    pub async fn stop(&mut self) {
        self.credentials = None;
    }

    async fn validate_api_key(&self, key: &str) -> bool {
        let url = format!("{}/models", self.client.base_url());
        let result = reqwest::Client::new()
            .get(&url)
            .header("Authorization", format!("Bearer {}", key))
            .timeout(std::time::Duration::from_secs(15))
            .send()
            .await;
        matches!(result, Ok(resp) if resp.status().is_success())
    }

    /// Perform device-based auto-signup.
    pub async fn authenticate_with_device(&mut self) -> Result<CloudCredentials> {
        let device_id = derive_device_id();
        let platform = detect_platform();
        let app_version = std::env::var("ELIZAOS_CLOUD_APP_VERSION")
            .unwrap_or_else(|_| "2.0.0-alpha".to_string());

        info!("[CloudAuth] Authenticating device (platform={})", platform);

        let body = serde_json::json!({
            "deviceId": device_id,
            "platform": platform,
            "appVersion": app_version,
            "deviceName": hostname::get()
                .ok()
                .and_then(|h| h.into_string().ok())
                .unwrap_or_else(|| "unknown".to_string()),
        });

        let resp = self.client.post_unauthenticated("/device-auth", &body).await?;
        let data = resp
            .get("data")
            .ok_or_else(|| ElizaCloudError::invalid_request("Missing data in response", vec![]))?;

        let api_key = data["apiKey"].as_str().unwrap_or_default().to_string();
        let user_id = data["userId"].as_str().unwrap_or_default().to_string();
        let org_id = data["organizationId"].as_str().unwrap_or_default().to_string();
        let credits = data["credits"].as_f64().unwrap_or(0.0);
        let is_new = data["isNew"].as_bool().unwrap_or(false);

        self.client.set_api_key(&api_key);

        let creds = CloudCredentials {
            api_key,
            user_id,
            organization_id: org_id,
            authenticated_at: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap_or_default()
                .as_secs_f64(),
        };
        self.credentials = Some(creds.clone());

        let action = if is_new { "New account created" } else { "Authenticated" };
        info!("[CloudAuth] {} (credits: ${:.2})", action, credits);

        Ok(creds)
    }

    pub fn is_authenticated(&self) -> bool {
        self.credentials.is_some()
    }

    pub fn credentials(&self) -> Option<&CloudCredentials> {
        self.credentials.as_ref()
    }

    pub fn api_key(&self) -> Option<&str> {
        self.credentials
            .as_ref()
            .map(|c| c.api_key.as_str())
            .or_else(|| self.client.api_key())
    }

    pub fn client(&self) -> &CloudApiClient {
        &self.client
    }

    pub fn client_mut(&mut self) -> &mut CloudApiClient {
        &mut self.client
    }

    pub fn user_id(&self) -> Option<&str> {
        self.credentials.as_ref().map(|c| c.user_id.as_str())
    }

    pub fn organization_id(&self) -> Option<&str> {
        self.credentials.as_ref().map(|c| c.organization_id.as_str())
    }
}

fn derive_device_id() -> String {
    use std::collections::hash_map::DefaultHasher;
    use std::hash::{Hash, Hasher};

    let hostname = hostname::get()
        .ok()
        .and_then(|h| h.into_string().ok())
        .unwrap_or_else(|| "unknown".to_string());

    let mut hasher = DefaultHasher::new();
    hostname.hash(&mut hasher);
    std::env::consts::OS.hash(&mut hasher);
    std::env::consts::ARCH.hash(&mut hasher);
    format!("{:016x}", hasher.finish())
}

fn detect_platform() -> &'static str {
    match std::env::consts::OS {
        "macos" => "macos",
        "windows" => "windows",
        "linux" => "linux",
        _ => "linux",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_derive_device_id_deterministic() {
        let id1 = derive_device_id();
        let id2 = derive_device_id();
        assert_eq!(id1, id2);
        assert!(!id1.is_empty());
    }

    #[test]
    fn test_detect_platform() {
        let plat = detect_platform();
        assert!(["macos", "windows", "linux"].contains(&plat));
    }

    #[test]
    fn test_new_service_not_authenticated() {
        let svc = CloudAuthService::new().unwrap();
        assert!(!svc.is_authenticated());
        assert!(svc.credentials().is_none());
        assert!(svc.user_id().is_none());
    }
}
